File: cpp/kmc/distribution.cpp
//
//  distribution.cpp
//  kmc
//
//  Created by Timothy Palpant on 3/31/13.
//  Copyright (c) 2013 Timothy Palpant. All rights reserved.
//

#include "distribution.h"

namespace kmc {
  namespace plugin {
    void Distribution::configure(const boost::property_tree::ptree& pt) {
      p_ = pt.get<boost::filesystem::path>("output");
      std::string particle = pt.get<std::string>("state");
      state_ = lattice::State::for_name(particle);
    }
    
    void Distribution::boot(kmc::lattice::Lattice* lattice) {
      lattice_ = lattice;
      last_state_ = std::vector<lattice::State*>(lattice_->states());
      dist_ = std::vector<double>(lattice_->size());
      of_.open(p_.c_str());
      of_ << "# Position\tProbability" << std::endl;
    }
    
    void Distribution::process(double time) {
      double dt = time - last_time_;
      for (std::size_t i = 0; i < last_state_.size(); i++) {
        if (last_state_[i] == state_) {
          dist_[i] += dt;
        }
        last_state_[i] = lattice_->get(i);
      }
      last_time_ = time;
    }
    
    void Distribution::close() {
      for (std::size_t i = 0; i < dist_.size(); i++) {
        of_ << i << '\t' << dist_[i]/last_time_ << std::endl;
      }
      of_.close();
    }
  }
}

File: java/src/us/palpant/science/kmc/plugins/Distribution.java
package us.palpant.science.kmc.plugins;

import java.io.IOException;
import java.io.PrintWriter;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import org.apache.log4j.Logger;

import us.palpant.Ark;
import us.palpant.science.kmc.State;
import us.palpant.science.kmc.geometry.Lattice;

/**
 * Calculate the fraction of the time that each lattice point
 * is in a certain state
 * 
 * @author timpalpant
 *
 */
public class Distribution extends Plugin {
  
  private static final Logger log = Logger.getLogger(Distribution.class);
  
  private final State state;
  private final Path outputFile;
  private double lastTime = 0;
  private Lattice lastState;
  private double[] dist;
  
  public Distribution(Lattice lattice, Ark config) {
    super(lattice);
    lastState = new Lattice(lattice.size(), lattice.getBoundaryCondition());
    dist = new double[lattice.size()];
    state = State.forName((String)config.get("state"));
    outputFile = Paths.get((String)config.get("name"));
  }

  @Override
  public void process(double time) {
    double dt = time - lastTime;
    for (int i = 0; i < lastState.size(); i++) {
      if (lastState.get(i) == state) {
        dist[i] += dt;
      }
      lastState.set(i, lattice.get(i));
    }
    
    lastTime = time;
  }

  @Override
  public void close() throws IOException {
    log.debug("Writing distribution to output file: "+outputFile);
    try (PrintWriter writer = new PrintWriter(Files.newBufferedWriter(outputFile, Charset.defaultCharset()))) {
      for (int i = 0; i < lastState.size(); i++) {
        writer.println(i+"\t"+dist[i]/lastTime);
      }
    }
  }

}
